File: Problems/03-Balanced-Binary-Tree/Balanced-Binary-Tree.cpp
#include <bits/stdc++.h>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	bool check(TreeNode* root, int& height)
	{
		if (root == nullptr) {
			height = -1;
			return true;
		}
		int leftHeight{};
		int rightHeight{};
		bool balanced = check(root->left, leftHeight) && check(root->right, rightHeight) && (abs(leftHeight - rightHeight) <= 1);
		height = max(leftHeight, rightHeight) + 1;
		return balanced
	}
    bool isBalanced(TreeNode* root) {
    	int height{};
    	return check(root, height);
    }
};

File: Problems/03-Balanced-Binary-Tree/Balanced-Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root) -> bool:
        n = self.check(root)
        if n==-1:
            return False
        else:
            return True


    def check(self,root):
        if root is None:
            return 0
        lh= self.check(root.left)
        rh= self.check(root.right)
        if lh==-1 or rh ==-1 or abs(lh-rh)>1:
            return -1
        else:
            return 1+max(lh,rh)